### ISSUE TITLE
Type narrowing of condition leads to unexpected error for assignment

### ISSUE THREAD

**Issue by krisztianb** [microsoft/TypeScript#60014]
### 🔎 Search Terms

type narrowing, Type 'number' is not assignable to type '-1', constant array

### 🕗 Version & Regression Information

- This changed between versions 5.4.5 and 5.5.0

### ⏯ Playground Link

https://www.typescriptlang.org/play/?#code/MYewdgzgLgBANgUwG4LgeQE4BMEZgXhgG0BaARgBoZyqbrL7aG6WyBdAbgCgBLAMxgAKRCnTZcAOmS4AnoMFIAhnACuCAJQEAfDCWqEBfIXLrNAby4wr8aWJwYiABjYEYZbgF8gA

### 💻 Code

```ts
const levelOrder = [-1, -1, -1, -1, -1, -1, -1, -1, -1];
if (levelOrder.every((value) => value === -1)) {
    levelOrder[0] = 1;
}
```

### 🙁 Actual behavior

The assignment gives the following TS error:

> Type 'number' is not assignable to type '-1'.ts(2322)
> const levelOrder: -1[]

The element in the array cannot be overwritten.

### 🙂 Expected behavior

The element in the array can be overwritten by any number.

### Additional information about the issue

Maybe related to this new feature: https://devblogs.microsoft.com/typescript/announcing-typescript-5-5/#control-flow-narrowing-for-constant-indexed-accesses

**Comment by MartinJohns** [comment]
IMO this is intentional and an unfortunate side-effect of the feature introduced in 5.5 (the one you linked) combined with the existence of the `every()` type guard overload.

You can simply add a type annotation to prevent the type-guard overload from being chosen:  
`levelOrder.every((value): boolean => value === -1)`

**Comment by jcalz** [comment]
Isn't it [inferred type predicates](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-5.html#inferred-type-predicates) and not the feature linked above?

**Comment by krisztianb** [comment]
Thank you for the replies and suggestions to adjust my code. If this works as intended I'm going to close this.

**Comment by RyanCavanaugh** [comment]
> This is inconsistent. TS determines assignability according to declaration type

This would be relevant if the code was
```ts
let levelOrder = [-1, -1, -1, -1, -1, -1, -1, -1, -1];
if (levelOrder.every((value) => value === -1)) {
    levelOrder = [1];
}
```
but this is a *property write*; if property writes used the un-narrowed type then you could easily corrupt a discriminated union

**Comment by Rudxain** [comment]
> if property writes used the un-narrowed type then you could easily corrupt a discriminated union

Ooh ok, thanks for explaining!
